lib/clientv2/resources.go

```diff
@@ -0,0 +1,281 @@
+func (c *resources) Get(opts options.GetOptions, kind, ns, name string) (resource, error) {
+	key := model.ResourceKey{
+		Kind:      kind,
+		Name:      name,
+		Namespace: ns,
+	}
+	kvp, err := c.backend.Get(key, opts.ResourceVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := c.kvPairToResource(kvp)
+	return out, nil
+}
+
+// List lists a resource from the backend datastore.
+func (c *resources) List(opts options.ListOptions, kind, listKind, ns, name string, listObj resourceList) error {
+	key := model.ResourceListOptions{
+		Kind:      kind,
+		Name:      name,
+		Namespace: ns,
+	}
+
+	// Query the backend.
+	kvps, err := c.backend.List(key, opts.ResourceVersion)
+	if err != nil {
+		return err
+	}
+
+	// Convert the slice of KVPairs to a slice of Objects.
+	resources := []runtime.Object{}
+	for _, kvp := range kvps.KVPairs {
+		resources = append(resources, c.kvPairToResource(kvp))
+	}
+	err = meta.SetList(listObj, resources)
+	if err != nil {
+		return err
+	}
+
+	// Finally, set the resource version and api group version of the list object.
+	listObj.GetListMeta().SetResourceVersion(kvps.Revision)
+	listObj.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   apiv2.Group,
+		Version: apiv2.VersionCurrent,
+		Kind:    listKind,
+	})
+
+	return nil
+}
+
+// Watch watches a specific resource or resource type.
+func (c *resources) Watch(opts options.ListOptions, kind, ns, name string) (watch.Interface, error) {
+	panic("Not implemented")
+	return nil, nil
+}
+
+// resourceToKVPair converts the resource to a KVPair that can be consumed by the
+// backend datastore client.
+func (c *resources) resourceToKVPair(opts options.SetOptions, kind string, in resource) *model.KVPair {
+	// Prepare the resource to remove non-persisted fields.
+	rv := in.GetObjectMeta().GetResourceVersion()
+	in.GetObjectMeta().SetResourceVersion("")
+	in.GetObjectMeta().SetSelfLink("")
+
+	// Make sure the kind and version are set before storing.
+	in.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   apiv2.Group,
+		Version: apiv2.VersionCurrent,
+		Kind:    kind,
+	})
+
+	// Create a KVPair using the "generic" resource Key, and the actual object as
+	// the value.
+	return &model.KVPair{
+		TTL:   opts.TTL,
+		Value: in,
+		Key: model.ResourceKey{
+			Kind:      kind,
+			Name:      in.GetObjectMeta().GetName(),
+			Namespace: in.GetObjectMeta().GetNamespace(),
+		},
+		Revision: rv,
+	}
+}
+
+// kvPairToResource converts a KVPair returned by the backend datastore client to a
+// resource.
+func (c *resources) kvPairToResource(kvp *model.KVPair) resource {
+	// Extract the resource from the returned value - the backend will already have
+	// decoded it.
+	out := kvp.Value.(resource)
+
+	// Remove the SelfLink which Calico does not use, and set the ResourceVersion from the
+	// value returned from the backend datastore.
+	out.GetObjectMeta().SetSelfLink("")
+	out.GetObjectMeta().SetResourceVersion(kvp.Revision)
```

Comment:
do we want to expose `creationTimestamp` to the clients? Not too worried about Get calls listing it but not sure if we want to let clients set it

Reply:
Why?  It's a valid field in the Metadata.  Why should it not be exposed - who uses it otherwise?

Reply:
The resource version is special because the source of truth is the datastore itself.  It provides the revision information and it is not stored in the resource data.

The self link is also a special field because it is filled in by the AAPIS and doesn't mean anything when just using the libcalico-go API.  We shouldn't persist it.  If it turns out the AAPIS needs it to be persisted we'll revisit, but I don't believe it does and I don't want to store data that we don't need to persist.